src/getRecords.ts

```diff
@@ -0,0 +1,57 @@
+// IDEA: getRecords handles cases of:
+// - getting all records from a table
+// - getting specific records from a table (single or multiple)
+// the difference is handled by either supplying a set of record id's or not
+// (and other params like filterByFormula etc)
+
+// also - can either call this once the airtable singleton has been set up (in which case "airtable" is passed to the function under the hood), or call it by itself and pass the api key explicitly
+
+import AirtableSDK from "airtable"
+
+import { RecordZ } from "./types/record"
+import { SchemaZodT } from "./types/schema"
+import { SelectQueryParamsT } from "./types/queryParams"
+
+export interface GetRecords<T extends SchemaZodT> {
+	tableId: string
+	baseId: string
+	schema: T
+	query?: SelectQueryParamsT
+}
+
+export interface GetRecordsWithAirtable<T extends SchemaZodT> extends GetRecords<T> {
+	airtable: AirtableSDK
+}
+export interface GetRecordsWithApiKey<T extends SchemaZodT> extends GetRecords<T> {
+	apiKey: string
+}
+
+export function getRecords<T extends SchemaZodT> ({
+	tableId,
+	schema,
+	baseId,
+	query,
+	...otherArgs
+}: GetRecordsWithAirtable<T> | GetRecordsWithApiKey<T>) {
+	let airtable
+	if ("airtable" in otherArgs) {
+		airtable = otherArgs.airtable
+	} else {
+		airtable = new AirtableSDK({ apiKey: otherArgs.apiKey })
+	}
+
+	// we overwrite the base RecordZ fields with the specific schema provided
+	const recordWithSchema = RecordZ.extend({
+		fields: schema
+	})
+
+	return airtable.base(baseId).table(tableId).select(query).all()
+	.then((records) => {
+		return Promise.all(
+			records.map((r) => {
+				// safeParseAsync in case of async refinements or transforms in schema, see https://zod.dev/?id=parseasync
+				return recordWithSchema.safeParseAsync(r)
+			})
+		)
+	})
```

Comment:
In principle I am all good with just returning the result of `safeParse`

But this leaves us without a strategy for how we deal with errors.

A problem I faced and something that is still unresolved in AFRA is what to do when you have errors in integrations (worst were batch operations).

To better understand how things work out the box I just spent maybe 15m going into the `airtable.js` sourcecode.

This bit was really interesting
https://github.com/Airtable/airtable.js/blob/899adb414ebc789aa3b0dcfe6c19113377315564/src/base.ts#L77

 - It shows how they do retries to make sure errors are not i/o errors
 - it shows the logic of how they resolve most errors as promise rejections with objects inside of them. [ie they don't `throw` or return `Errors`](https://github.com/Airtable/airtable.js/blob/master/src/airtable_error.ts#L1). Rather they return an error-object.

Now that I think about it, I am noticing it comes back to the question previously raised of: **what is the scope of this bamboo lib**. 

Currently the code above is a true wrapper around the airtable SDK. ie whatever Errors or error-objects are just returned to the consuming code. 

On one hand this is clean and an easy level of abstraction to find. On the other I see an opportunity to do more then this and provide devs with a more standardised approach to dealing with logic around failure.

But now thinking about this we could just do it via examples in the readme or tests.

Another example of something that could be shown in a test or in the Readme is the thing I worked out for filtering via a `.flatMap` for narrowing the types of the return.